pkg/reconciler/namespace/scheduler.go

```diff
@@ -0,0 +1,139 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namespace
+
+import (
+	"math/rand"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/clusters"
+	"k8s.io/klog/v2"
+
+	clusterv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/cluster/v1alpha1"
+	"github.com/kcp-dev/kcp/third_party/conditions/util/conditions"
+)
+
+type getClusterFunc func(name string) (*clusterv1alpha1.Cluster, error)
+type listClustersFunc func(selector labels.Selector) ([]*clusterv1alpha1.Cluster, error)
+
+type namespaceScheduler struct {
+	getCluster   getClusterFunc
+	listClusters listClustersFunc
+}
+
+// AssignCluster returns the name of the cluster to assign to the provided
+// namespace. The current cluster assignment will be returned if it is valid or if
+// the automatic scheduling is disabled for the namespace. An new assignment will
+// be attempted if the current assignment is empty or invalid.
+func (s *namespaceScheduler) AssignCluster(ns *corev1.Namespace) (string, error) {
+	assignedCluster := ns.Labels[clusterLabel]
+
+	schedulingDisabled := !scheduleRequirement.Matches(labels.Set(ns.Labels))
+	if schedulingDisabled {
+		klog.Infof("Automatic scheduling is disabled for namespace %s|%s", ns.ClusterName, ns.Name)
+		return assignedCluster, nil
+	}
+
+	if assignedCluster != "" {
+		isValid, err := s.isValidCluster(ns.ClusterName, assignedCluster)
+		if err != nil {
+			return "", err
+		}
+		if isValid {
+			return assignedCluster, nil
+		}
+		// Currently assigned cluster no longer exists or is not ready.
+	}
+
+	allClusters, err := s.listClusters(labels.Everything())
+	if err != nil {
+		return "", err
+	}
+	return pickCluster(allClusters, ns.ClusterName), nil
+}
+
+// isValidCluster checks whether the given cluster name exists and is valid for
+// the purposes of any namespace already scheduled to it (i.e., if it reports
+// as Ready, and any evictAfter value, if specified, has not yet passed).
+//
+// It doesn't take into account Unschedulable, and should only be used when
+// determining if a cluster that a namespace has already been assigned to
+// should keep having that namespace.
+func (s *namespaceScheduler) isValidCluster(lclusterName, clusterName string) (bool, error) {
+	cluster, err := s.getCluster(clusters.ToClusterAwareKey(lclusterName, clusterName))
+	if apierrors.IsNotFound(err) {
+		klog.Infof("Cluster %q|%q does not exist", lclusterName, clusterName)
```

Comment:
nit: logs in predicates or getters have smell

Reply:
make the func return `bool, message string, error`

Reply:
done